### PR TITLE
ddns-updater: Add module and update script

### PR DESCRIPTION
**Pull request by delliottxyz**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by delliottxyz** [comment]
Merge conflict was trivial collision in release note, now fixed.

**Comment by delliottxyz** [comment]
Merge conflict again introduced in manual and now resolved

**Comment by delliottxyz** [comment]
Updated to latest version of ddns-updater, 2.7.0 and updated release notes to 24.11 instead of 24.05

File: nixos/modules/services/networking/ddns-updater.nix [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+      default = {
+        LISTENING_ADDRESS = ":8000";
+        DATADIR = "/updater/data";
+      };
+    };
+
+    user = lib.mkOption {
+      type = lib.types.str;
+      description = "User to run service as. Owns DATADIR";
+      default = "ddns-updater";
```
Comment (NickCao) [inline nixos/modules/services/networking/ddns-updater.nix:29]: Prefer using `DynamicUser` and `StateDirectory` if possible.

File: nixos/modules/services/networking/ddns-updater.nix [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    services.ddns-updater.environmentalVariables = {
+      LISTENING_ADDRESS = lib.mkDefault ":8000";
+      DATADIR = lib.mkDefault "/updater/data";
+    };
+
+    systemd.services.ddns-updater = {
+      wantedBy = [ "multi-user.target" ];
+      unitConfig = {
+        Description = "DDNS-updater service";
+        After = [ "network.target" ];
```
Comment (NickCao) [inline nixos/modules/services/networking/ddns-updater.nix:43]: `network-online.target` might be more appropriate.

File: nixos/modules/services/networking/ddns-updater.nix [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      };
+      serviceConfig = {
+        TimeoutSec = "5min";
+        Environment =
+          let
+            envVars = lib.attrsets.mapAttrsToList (name: value: name + "=" + value) cfg.environmentalVariables;
+          in
+          envVars;
```
Comment (NickCao) [inline nixos/modules/services/networking/ddns-updater.nix:51]: Please use the `systemd.services.<name>.environment` option.

File: nixos/modules/services/networking/ddns-updater.nix [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        ExecStart = "${cfg.package}/bin/ddns-updater";
```
Comment (NickCao) [inline nixos/modules/services/networking/ddns-updater.nix:52]:
```suggestion
        ExecStart = lib.getExe cfg.package;
```

**Review by NickCao — COMMENTED** [review]
Please split the version bump and the module addition to their own commits.

**Comment by delliottxyz** [comment]
Hey, thanks for the review!
I'm away from my laptop for a few days but I'll get to this when I'm back to it. 
Thanks again.

**Comment by delliottxyz** [comment]
Hey, 
I believe I have made the requested changes, and resolved the latest manual-induced merge conflict.

File: nixos/modules/services/networking/ddns-updater.nix [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    };
+
+    systemd.services.ddns-updater = {
+      wantedBy = [ "multi-user.target" ];
+      environment = cfg.environment // {
+        DATADIR = "/var/lib/ddns-updater";
+      };
+      unitConfig = {
+        Description = "DDNS-updater service";
+        After = [ "network-online.target" ];
```
Comment (NickCao) [inline nixos/modules/services/networking/ddns-updater.nix:39]:
```suggestion
```

File: nixos/modules/services/networking/ddns-updater.nix [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.ddns-updater.environment = {
+      LISTENING_ADDRESS = lib.mkDefault defaultEnv.LISTENING_ADDRESS;
+    };
+
+    systemd.services.ddns-updater = {
+      wantedBy = [ "multi-user.target" ];
```
Comment (NickCao) [inline nixos/modules/services/networking/ddns-updater.nix:33]:
```suggestion
      wantedBy = [ "multi-user.target" ];
      wants = [ "network-online.target" ];
      after = [ "network-online.target" ];
```

File: nixos/modules/services/networking/ddns-updater.nix [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+      environment = cfg.environment // {
+        DATADIR = "/var/lib/ddns-updater";
```
Comment (NickCao) [inline nixos/modules/services/networking/ddns-updater.nix:35]:
```suggestion
        DATADIR = "%S/ddns-updater";
```

See https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#Specifiers

File: nixos/modules/services/networking/ddns-updater.nix [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+{
+  options.services.ddns-updater = {
+    enable = lib.mkEnableOption "Container to update DNS records periodically with WebUI for many DNS providers";
+
+    package = lib.mkPackageOption pkgs "ddns-updater" { };
+
+    environment = lib.mkOption {
+      type = lib.types.attrsOf lib.types.str;
+      description = "Environment variables to be set for the ddns-updater service. DATADIR is ignored to enable using systemd DynamicUser. For full list see https://github.com/qdm12/ddns-updater";
+      default = defaultEnv;
```
Comment (NickCao) [inline nixos/modules/services/networking/ddns-updater.nix:23]:
```suggestion
      default = { };
      defaultText = lib.literalExpression ''
        {
          LISTENING_ADDRESS = ":8000";
        }
      '';
```

Default already set down below.

File: nixos/modules/services/networking/ddns-updater.nix [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.ddns-updater.environment = {
+      LISTENING_ADDRESS = lib.mkDefault defaultEnv.LISTENING_ADDRESS;
```
Comment (NickCao) [inline nixos/modules/services/networking/ddns-updater.nix:29]:
```suggestion
      LISTENING_ADDRESS = lib.mkDefault ":8000";
```

File: nixos/modules/services/networking/ddns-updater.nix [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  cfg = config.services.ddns-updater;
+  defaultEnv = {
+    LISTENING_ADDRESS = ":8000";
+  };
```
Comment (NickCao) [inline nixos/modules/services/networking/ddns-updater.nix:12]:
```suggestion
```

**Comment by delliottxyz** [comment]
I have implemented the requested changes, and realized that there was no reason to have a default listening address at all and so removed all default envvars.

Thanks for the time you're putting into reviewing this!

**Comment by delliottxyz** [comment]
I have implemented the requested changes, and realized that there was no reason to have a default listening address at all and so removed all default envvars.

Thanks for the time you're putting into reviewing this!

**Comment by NickCao** [comment]
@ofborg test ddns-updater

**Review by NickCao — COMMENTED** [review]
Diff LGTM, thanks for your hard work! And a final nitpick: could you link the nixos test into the package like in https://github.com/NixOS/nixpkgs/blob/34bd4321d43b88fea95642c287773db0d9f494e5/pkgs/applications/radio/cloudlog/default.nix#L32, this would allow ofborg to automatically pick it up.

**Review by NickCao — APPROVED** [review]
Waiting for ofborg.

**Comment by delliottxyz** [comment]
Thanks for your work @NickCao !

**Pull request merged**